modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,449 @@
+     * to this storage to specified position {@code pos}.
+     *
+     * @param pos Pointer to a position.
+     * @param inpBuf Input byte array.
+     * @param inpOff Start offset in the input array.
+     * @param inpLen Number of bytes to write.
+     */
+    void write(int pos, byte[] inpBuf, int inpOff, int inpLen) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < inpLen)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        Objects.checkFromIndexSize(inpOff, inpLen, inpBuf.length);
+
+        int newLen = Math.max(pos + inpLen, len);
+
+        reallocateIfRequired(newLen);
+
+        U.arrayCopy(inpBuf, inpOff, arr, pos, inpLen);
+
+        len = newLen;
+    }
+
+    /**
+     * Read a byte from this buffer from specified position {@code pos}.
+     *
+     * @param pos Position.
+     * @return Byte read from the Blob. -1 if end of data reached.
+     */
+    int read(int pos) {
+        if (pos >= len)
+            return -1;
+
+        return arr[pos + off] & 0xff;
+    }
+
+    /**
+     * Write a byte to this buffer to specified position {@code pos}.
+     *
+     * <p>The byte to be written is the eight low-order bits of the
+     * argument {@code b}. The 24 high-order bits of {@code b}b are ignored.
+     *
+     * @param pos Pointer to a position.
+     * @param b Byte to write.
+     */
+    void write(int pos, int b) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < 1)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        int newLen = Math.max(pos + 1, len);
+
+        reallocateIfRequired(newLen);
+
+        arr[pos] = (byte)b;
+
+        len = newLen;
+    }
+
+    /**
+     * Ensure capacity.
+     *
+     * @param newLen The new data length the buffer should be able to hold.
+     */
+    private void reallocateIfRequired(int newLen) {
+        if (newLen - arr.length > 0 || isReadOnly)
+            reallocate(capacity(arr.length, newLen));
+    }
+
+    /**
+     * Calculate new capacity.
+     *
+     * @param cap Current capacity.
+     * @param reqCap Required new capacity.
+     * @return New capacity.
+     */
+    protected static int capacity(int cap, int reqCap) {
+        if (reqCap <= MIN_CAP)
+            return MIN_CAP;
+
+        int resCap = Math.max(cap, MIN_CAP);
+
+        while (resCap < reqCap) {
+            resCap <<= 1;
+
+            if (resCap < 0)
+                return MAX_ARRAY_SIZE;
+        }
+
+        return resCap;
+    }
+
+    /**
+     * Allocate the new underlining array and copy data.
+     *
+     * @param newCapacity New capacity.
+     */
+    private void reallocate(int newCapacity) {
+        byte[] newBuf = new byte[newCapacity];
+
+        U.arrayCopy(arr, off, newBuf, 0, len);
+
+        arr = newBuf;
+        off = 0;
+        isReadOnly = false;
```

Comment:
We must change `len` here.

Reply:
`newLen`

Reply:
Sorry, but there is no such variable.

Reply:
I mean `newCapacity`.

Reply:
newCapacity is a length of the new byte[] internal buffer.

len is a length of actual data in this buffer.

Why do you think one should be changed to value of another?

Reply:
Because, in the all cases except the one with truncate it's the same value

Reply:
No.  Usually buffer is larger then data.  We allocate more space in advance to support streaming way of operations - we expect that more data would be written very very soon.

Reply:
Two places where we call reallocationIfRequired:

```
        reallocateIfRequired(newLen);

        U.arrayCopy(inpBuf, inpOff, arr, pos, inpLen);

        len = newLen;
...

        reallocateIfRequired(newLen);

        arr[pos] = (byte)b;

        len = newLen;

```

Both of them change `len` to the parameter of the `reallocateIfRequired`.

Third case when we call reallocation directly:

```
        this.len = len;

        reallocate(Math.max(MIN_CAP, len));
```

We also change `len` value to the parameter of `reallocate` method.
So, we can move change of len to the one of methods.

From the code correction point of view - we must change buffere parameters in the fewest possible places.
Looks like we must add one more parameter to the `reallocate` method - `len`

Reply:
Okay, if you want to. But IMHO it's wrong.  

`reallocate` and  `reallocateIfRequired` do now exactly what names suggest - reallocate internal buffer. But do not change the buffer state.  It's callers who know how to change the buffer state (deal with `len` in particular).